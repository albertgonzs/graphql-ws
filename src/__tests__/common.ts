import { validateMessage, MessageType } from '../common';

it('should report invalid messages with descriptive errors', () => {
  for (const invalidMessage of [
    // straight up invalid
    {},
    '',
    [],
    0,
    9,
    Symbol,
    Object,
    () => {
      /**/
    },

    // invalid 'type' prop
    {
      type: '',
    },
    {
      type: undefined,
    },
    {
      type: 0,
    },
    {
      type: 'nuxt',
    },

    // invalid connection_init, connection_ack, ping and pong message
    {
      type: MessageType.ConnectionInit,
      payload: '',
    },
    {
      type: MessageType.ConnectionInit,
      payload: 0,
    },
    {
      type: MessageType.ConnectionInit,
      payload: undefined,
    },
    {
      type: MessageType.ConnectionAck,
      payload: '',
    },
    {
      type: MessageType.Ping,
      payload: 0,
    },
    {
      type: MessageType.Pong,
      payload: undefined,
    },

    // invalid subscribe message
    {
      type: MessageType.Subscribe,
    },
    {
      id: 0,
      type: MessageType.Subscribe,
    },
    {
      id: '',
      type: MessageType.Subscribe,
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: [],
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: '',
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {},
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        query: 0,
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        query: {},
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        operationName: 0,
        query: '',
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        operationName: {},
        query: '',
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        query: '',
        variables: '',
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        query: '',
        extensions: '',
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        query: '',
        extensions: 0,
      },
    },

    // invalid next message
    {
      type: MessageType.Next,
    },
    {
      id: undefined,
      type: MessageType.Next,
    },
    {
      id: '',
      type: MessageType.Next,
    },
    {
      id: 'id',
      type: MessageType.Next,
    },
    {
      id: 'id',
      type: MessageType.Next,
      payload: '',
    },

    // invalid error message
    {
      type: MessageType.Error,
    },
    {
      id: '',
      type: MessageType.Error,
    },
    {
      id: 'id',
      type: MessageType.Error,
    },
    {
      id: 'id',
      type: MessageType.Error,
      payload: {},
    },
    {
      id: 'id',
      type: MessageType.Error,
      payload: '',
    },
    {
      id: 'id',
      type: MessageType.Error,
      payload: [],
    },
    {
      id: 'id',
      type: MessageType.Error,
      payload: [{ iam: 'invalid' }],
    },

    // invalid complete message
    {
      type: MessageType.Complete,
    },
    {
      id: '',
      type: MessageType.Complete,
    },
    {
      id: 0,
      type: MessageType.Complete,
    },
  ]) {
    expect(() =>
      validateMessage(invalidMessage),
    ).toThrowErrorMatchingSnapshot();
  }
});

it('should accept valid messages', () => {
  for (const valid of [
    // valid connection_init, connection_ack, ping and pong message
    {
      type: MessageType.ConnectionInit,
    },
    {
      type: MessageType.ConnectionInit,
      payload: {},
    },
    {
      type: MessageType.ConnectionAck,
    },
    {
      type: MessageType.ConnectionAck,
      payload: {},
    },
    {
      type: MessageType.Ping,
    },
    {
      type: MessageType.Ping,
      payload: {},
    },
    {
      type: MessageType.Pong,
    },
    {
      type: MessageType.Pong,
      payload: {},
    },

    // valid subscribe message
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        query: '',
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        operationName: undefined,
        variables: undefined,
        extensions: undefined,
        query: '',
      },
    },
    {
      id: 'id',
      type: MessageType.Subscribe,
      payload: {
        operationName: null,
        variables: null,
        extensions: null,
        query: '',
      },
    },

    // valid error message
    {
      id: 'id',
      type: MessageType.Error,
      payload: [{ message: 'I am Error' }],
    },

    // valid complete message
    {
      id: 'id',
      type: MessageType.Complete,
    },
  ]) {
    expect(() => validateMessage(valid)).not.toThrow();
  }
});
