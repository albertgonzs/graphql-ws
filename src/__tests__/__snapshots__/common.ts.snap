// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should report invalid messages with descriptive errors 1`] = `"Message is missing the 'type' property"`;

exports[`should report invalid messages with descriptive errors 2`] = `"Message is expected to be an object, but got string"`;

exports[`should report invalid messages with descriptive errors 3`] = `"Message is expected to be an object, but got array"`;

exports[`should report invalid messages with descriptive errors 4`] = `"Message is expected to be an object, but got number"`;

exports[`should report invalid messages with descriptive errors 5`] = `"Message is expected to be an object, but got number"`;

exports[`should report invalid messages with descriptive errors 6`] = `"Message is expected to be an object, but got function"`;

exports[`should report invalid messages with descriptive errors 7`] = `"Message is expected to be an object, but got function"`;

exports[`should report invalid messages with descriptive errors 8`] = `"Message is expected to be an object, but got function"`;

exports[`should report invalid messages with descriptive errors 9`] = `"Message is missing the 'type' property"`;

exports[`should report invalid messages with descriptive errors 10`] = `"Message is missing the 'type' property"`;

exports[`should report invalid messages with descriptive errors 11`] = `"Message is missing the 'type' property"`;

exports[`should report invalid messages with descriptive errors 12`] = `"Invalid message 'type' property "nuxt""`;

exports[`should report invalid messages with descriptive errors 13`] = `""connection_init" message expects the 'payload' property to be an object or missing, but got """`;

exports[`should report invalid messages with descriptive errors 14`] = `""connection_init" message expects the 'payload' property to be an object or missing, but got "0""`;

exports[`should report invalid messages with descriptive errors 15`] = `""connection_init" message expects the 'payload' property to be an object or missing, but got "undefined""`;

exports[`should report invalid messages with descriptive errors 16`] = `""connection_ack" message expects the 'payload' property to be an object or missing, but got """`;

exports[`should report invalid messages with descriptive errors 17`] = `""ping" message expects the 'payload' property to be an object or missing, but got "0""`;

exports[`should report invalid messages with descriptive errors 18`] = `""pong" message expects the 'payload' property to be an object or missing, but got "undefined""`;

exports[`should report invalid messages with descriptive errors 19`] = `""subscribe" message expects the 'id' property to be a string, but got undefined"`;

exports[`should report invalid messages with descriptive errors 20`] = `""subscribe" message expects the 'id' property to be a string, but got number"`;

exports[`should report invalid messages with descriptive errors 21`] = `""subscribe" message requires a non-empty 'id' property"`;

exports[`should report invalid messages with descriptive errors 22`] = `""subscribe" message expects the 'payload' property to be an object, but got undefined"`;

exports[`should report invalid messages with descriptive errors 23`] = `""subscribe" message expects the 'payload' property to be an object, but got array"`;

exports[`should report invalid messages with descriptive errors 24`] = `""subscribe" message expects the 'payload' property to be an object, but got string"`;

exports[`should report invalid messages with descriptive errors 25`] = `""subscribe" message payload expects the 'query' property to be a string, but got undefined"`;

exports[`should report invalid messages with descriptive errors 26`] = `""subscribe" message payload expects the 'query' property to be a string, but got number"`;

exports[`should report invalid messages with descriptive errors 27`] = `""subscribe" message payload expects the 'query' property to be a string, but got object"`;

exports[`should report invalid messages with descriptive errors 28`] = `""subscribe" message payload expects the 'operationName' property to be a string or nullish or missing, but got number"`;

exports[`should report invalid messages with descriptive errors 29`] = `""subscribe" message payload expects the 'operationName' property to be a string or nullish or missing, but got object"`;

exports[`should report invalid messages with descriptive errors 30`] = `""subscribe" message payload expects the 'variables' property to be a an object or nullish or missing, but got string"`;

exports[`should report invalid messages with descriptive errors 31`] = `""subscribe" message payload expects the 'extensions' property to be a an object or nullish or missing, but got string"`;

exports[`should report invalid messages with descriptive errors 32`] = `""subscribe" message payload expects the 'extensions' property to be a an object or nullish or missing, but got number"`;

exports[`should report invalid messages with descriptive errors 33`] = `""next" message expects the 'id' property to be a string, but got undefined"`;

exports[`should report invalid messages with descriptive errors 34`] = `""next" message expects the 'id' property to be a string, but got undefined"`;

exports[`should report invalid messages with descriptive errors 35`] = `""next" message requires a non-empty 'id' property"`;

exports[`should report invalid messages with descriptive errors 36`] = `""next" message expects the 'payload' property to be an object, but got undefined"`;

exports[`should report invalid messages with descriptive errors 37`] = `""next" message expects the 'payload' property to be an object, but got string"`;

exports[`should report invalid messages with descriptive errors 38`] = `""error" message expects the 'id' property to be a string, but got undefined"`;

exports[`should report invalid messages with descriptive errors 39`] = `""error" message requires a non-empty 'id' property"`;

exports[`should report invalid messages with descriptive errors 40`] = `""error" message expects the 'payload' property to be an array of GraphQL errors, but got undefined"`;

exports[`should report invalid messages with descriptive errors 41`] = `""error" message expects the 'payload' property to be an array of GraphQL errors, but got {}"`;

exports[`should report invalid messages with descriptive errors 42`] = `""error" message expects the 'payload' property to be an array of GraphQL errors, but got """`;

exports[`should report invalid messages with descriptive errors 43`] = `""error" message expects the 'payload' property to be an array of GraphQL errors, but got []"`;

exports[`should report invalid messages with descriptive errors 44`] = `""error" message expects the 'payload' property to be an array of GraphQL errors, but got [{"iam":"invalid"}]"`;

exports[`should report invalid messages with descriptive errors 45`] = `""complete" message expects the 'id' property to be a string, but got undefined"`;

exports[`should report invalid messages with descriptive errors 46`] = `""complete" message requires a non-empty 'id' property"`;

exports[`should report invalid messages with descriptive errors 47`] = `""complete" message expects the 'id' property to be a string, but got number"`;
